### DISCUSSION TITLE
Applying Main App Adaptive Mesh to Sub App in MultiApp

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#27057] [Q&A General] [answered]
Dear MOOSE experts,

I'm currently working on transfer adaptive mesh refinement from a main application into a sub-application within a MultiApp. We found that [LevelSetMeshRefinementTransfer](https://mooseframework.inl.gov/moose/source/transfers/LevelSetMeshRefinementTransfer.html) can meet such requirements, but it is located in the level set module instead of the moose module. Based on the provided reference calculation example ([parent.i](https://github.com/idaholab/moose/blob/next/modules/level_set/test/tests/transfers/markers/multi_level/parent.i)), I wrote my own [calculation example file](https://github.com/PengWei97/yinglong/commit/bd2fea708566d60097eaf983e9dcadf9c02e5999).

However, upon implementing a similar setup in our custom application, we've encountered an issue where the mesh adaptivity does not seem to be executed within the sub-application. Additionally, as shown in the attached image, 
![image](https://github.com/idaholab/moose/assets/56183775/231477a0-9a72-464c-a285-b33695e83304)
the markers intended for transfer do not appear to be successfully replicated in the sub-app.

I'm seeking guidance on the following points:

1. Could there be specific considerations or settings that I might have overlooked in applying LevelSetMeshRefinementTransfer to ensure successful mesh refinement and marker transfer to the sub-app?
2. Is the Level Set Module's LevelSetMeshRefinementTransfer inherently applicable to other apps, or are there strategies to generalize its application for use with other MOOSE-based Apps?

Any insights, suggestions, or references to similar cases would be immensely appreciated. Our goal is to effectively leverage MOOSE's capabilities for our project, and resolving this issue is crucial for our progress.

Thank you in advance for your support and guidance.
Best regards,
Wei

**Comment by GiudGiud** [comment]
Hello

Can you transfer the marker on timestep-begin? Then execute adaptivity during the time step.

I think the timing of execution is the challenging part here

Guillaume

**Reply by PengWei97** [reply]
Hi Guillaume,

Thank you for your suggestion regarding the timing of marker transfer and adaptivity execution. Following your advice, I attempted to configure the execution of the MultiApp to occur at both `timestep_begin` and `timestep_end` to see if it would facilitate the successful transfer of markers into the sub-app. Unfortunately, this adjustment did not lead to the markers being executed within the sub-application as intended.

Further investigation into the issue revealed some additional insights. I inserted breakpoints within the `LevelSetMeshRefinementTransfer` code to monitor its execution flow. Through this method, I observed that `LevelSetMeshRefinementTransfer::initialSetup()` is being called and executed as expected. However, `LevelSetMeshRefinementTransfer::execute()`, which is crucial for the actual transfer of adaptivity information, does not seem to be triggered at all. This observation contrasts with the behavior seen in the reference example (parent.i), where `LevelSetMeshRefinementTransfer::execute()` is executed effectively.

Moreover, I encountered a limitation regarding the `execution_on` parameter within the `LevelSetMeshRefinementTransfer`. It appears that this parameter is private to `LevelSetMeshRefinementTransfer` and thus cannot be explicitly set in the input file. 

Given these observations, it seems we are facing a more complex challenge related to the execution mechanism of the `LevelSetMeshRefinementTransfer`. Any further insights or recommendations on how to ensure the `execute()` method is triggered appropriately would be greatly appreciated.

Thank you once again for your support and for helping to steer the investigation in the right direction.

Best regards,

Wei

**Reply by GiudGiud** [reply]
Hello

I dont think we need the levelset mesh refinement transfer to transfer markers but I may be wrong. I thought I saw a different post last year transfering a marker with a regular transfer. Do you really want an exact synchronization of the two meshes or can they be offset by one time step?

> Moreover, I encountered a limitation regarding the execution_on parameter within the LevelSetMeshRefinementTransfer. It appears that this parameter is private to LevelSetMeshRefinementTransfer and thus cannot be explicitly set in the input file.

This may be. I can imagine we only use this transfer during rounds of adaptivity for the main app. I checked and the execution flags are hard set to:
```
{LevelSet::EXEC_COMPUTE_MARKERS, LevelSet::EXEC_ADAPT_MESH};
```

so to execute it, you need to make sure the adaptivity is active in the main app. Maybe force it to run with minimum iteration parameters if that's available?

Guillaume

**Reply by PengWei97** [reply]
Hi Guillaume,

Thank your very much for your enthusiastic help. Below is my reply to your suggestions,

> I thought I saw a different post last year transfering a marker with a regular transfer. 

You mean we use a regular method to transfer a marker and then run adaptivity in the sub-app, right? hugary1995 does this by directly transferring the variable used to evaluate the error in [ref1](https://groups.google.com/g/moose-users/c/5Bz7oTQU_6E/m/OCOy_FEQAwAJ) and [ref2](https://github.com/hugary1995/raccoon/blob/master/tutorials/adaptivity/elasticity.i).

> Do you really want an exact synchronization of the two meshes or can they be offset by one time step?

I believe  the correct approach is to initially run the main-app, followed by the transferring the mark information to the sub-app. Subsequently, we should transfer the relevent data to the sub-app, enabling the sub-app's mesh configuration aligns with that of the main-app. 

> so to execute it, you need to make sure the adaptivity is active in the main app. Maybe force it to run with minimum iteration parameters if that's available?

Yeah, I can confirm that the adaptivity mesh is executed at each time step in the main app, but `LevelSetMeshRefinementTransfer::execute()` is not found to be called in the sub-app. Also, in #18813, @aeslaughter said, "I created LevelSetMeshRefinementTransfer 5 years ago for this same concept. There might be a better way to do it now, but if not you can take a look at it for some inspiration." Perhaps he can provide more detailed advice.

In general, there are two idea. One is to directly transfer variables (e.g. bnds) based on hugary1995's methodl, and other is to use `LevelSetMeshRefinementTransfer`. I think the latter method is the best method, but there are still some issues to address. This necessitates further guidance for you and aeslaughter.

Wei

**Reply by GiudGiud** [reply]
I can look into why the example does not execute the transfer as expected.
Could you please reduce the [calculation example file](https://github.com/PengWei97/yinglong/commit/bd2fea708566d60097eaf983e9dcadf9c02e5999) down to a MOOSE-only minimal example?

**Reply by PengWei97** [reply]
Hi Guillaume, I have reduced the the [calculation example file](https://github.com/PengWei97/yinglong/commit/b67185ff546411605393f027b1ec7b81) down to a MOOSE-only minimal example.

**Reply by GiudGiud** [reply]
Thanks!
I m looking at the transfer but you should look at this error I hit using the devel version of the executable:
```


Assertion `_mesh.queryElemPtr(elem->id()) == elem' failed
Are you calling this method with a displaced mesh element?
at /Users/giudgl/projects/moose/framework/src/problems/FEProblemBase.C, line 1922
Stack frames: 24
0: 0   libmesh_devel.0.dylib               0x0000000110f762d0 libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char>>&) + 1044
1: 1   libmoose-devel.0.dylib              0x0000000106c35bdc FEProblemBase::reinitElemPhys(libMesh::Elem const*, std::__1::vector<libMesh::Point, std::__1::allocator<libMesh::Point>> const&, unsigned int) + 912
2: 2   libphase_field-devel.0.dylib        0x00000001095a45a8 GrainTracker::updateFieldInfo() + 1004
3: 3   libphase_field-devel.0.dylib        0x00000001095b72b0 GrainTracker::finalize() + 816
4: 4   libmoose-devel.0.dylib              0x0000000106bf76c4 FEProblemBase::joinAndFinalize(TheWarehouse::QueryCache<>, bool) + 556
5: 5   libmoose-devel.0.dylib              0x0000000106bf50bc FEProblemBase::computeUserObjectsInternal(MooseEnumItem const&, Moose::AuxGroup const&, TheWarehouse::QueryCache<>&) + 6044
6: 6   libmoose-devel.0.dylib              0x0000000106bf3670 FEProblemBase::computeUserObjects(MooseEnumItem const&, Moose::AuxGroup const&) + 232
7: 7   libmoose-devel.0.dylib              0x0000000106befe78 FEProblemBase::execute(MooseEnumItem const&) + 300
8: 8   libmoose-devel.0.dylib              0x000000010655c114 FixedPointSolve::solveStep(double&, double&, std::__1::set<unsigned long long, std::__1::less<unsigned long long>, std::__1::allocator<unsigned long long>> const&) + 860
9: 9   libmoose-devel.0.dylib              0x000000010655a304 FixedPointSolve::solve() + 3552
10: 10  libmoose-devel.0.dylib              0x000000010717926c TimeStepper::step() + 44
11: 11  libmoose-devel.0.dylib              0x000000010655f92c Transient::takeStep(double) + 180
12: 12  libmoose-devel.0.dylib              0x00000001070adf60 TransientMultiApp::solveStep(double, double, bool) + 2280
13: 13  libmoose-devel.0.dylib              0x0000000106bfe8c0 FEProblemBase::execMultiApps(MooseEnumItem, bool) + 1272
14: 14  libmoose-devel.0.dylib              0x000000010655c6d4 FixedPointSolve::solveStep(double&, double&, std::__1::set<unsigned long long, std::__1::less<unsigned long long>, std::__1::allocator<unsigned long long>> const&) + 2332
15: 15  libmoose-devel.0.dylib              0x000000010655a304 FixedPointSolve::solve() + 3552
16: 16  libmoose-devel.0.dylib              0x000000010717926c TimeStepper::step() + 44
17: 17  libmoose-devel.0.dylib              0x000000010655f92c Transient::takeStep(double) + 180
18: 18  libmoose-devel.0.dylib              0x000000010656acd8 Transient::execute() + 152
19: 19  libmoose-devel.0.dylib              0x00000001075c43f4 MooseApp::executeExecutioner() + 684
20: 20  libmoose-devel.0.dylib              0x00000001075bc9d8 MooseApp::run() + 4584
21: 21  combined-devel                      0x0000000102defbf8 void Moose::main<CombinedTestApp>(int, char**) + 152
22: 22  combined-devel                      0x0000000102defb54 main + 12
23: 23  dyld                                0x000000019a5ea0e0 start + 2360
[0] /Users/giudgl/projects/moose/framework/src/problems/FEProblemBase.C, line 1922, compiled Mar 11 2024 at 13:29:51
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1

```
I think there's some feature in GrainTracker that does not support displaced meshes that you are using with displaced meshes

**Reply by GiudGiud** [reply, marked as answer]
For the transfer, you have to add this
```
[Problem]
  type = LevelSetProblem
[]
```
to make the level set mesh refinement transfers work

**Reply by PengWei97** [reply]
Thank you for your guidance! 

The issue has been resolved. Adding this specified lines in main-app. I appreciate your help!

Besides,
> I think there's some feature in GrainTracker that does not support displaced meshes that you are using with displaced meshes

In fact, [GrainTrack](https://mooseframework.inl.gov/moose/source/postprocessors/GrainTracker.html) can support displaced meshes I think.

Wei
